File: cxx/rest/JsonGenerator.cpp
/**
 * @file JsonGenerator.cpp
 * @author Ronald T. Fernandez
 * @version 1.0
 */

#include "JsonGenerator.h"
#include "rapidjson/stringbuffer.h"
#include "rapidjson/prettywriter.h"

#include <iostream>

namespace rest {

void JsonGenerator::getJson(const std::string& root, const std::vector<std::string> elements, std::string& jsonText) {

	// Create a document
	rapidjson::Document document;
	document.SetObject();
	rapidjson::Document::AllocatorType& docAlloc = document.GetAllocator();

	// Add contents to the document
	rapidjson::Value list = rapidjson::Value(rapidjson::kArrayType);
	for (std::string element : elements) {
		rapidjson::Value v(element.c_str(), docAlloc);
		list.PushBack(v, docAlloc);
	}
	rapidjson::Value rootJson(root.c_str(), docAlloc);
	document.AddMember(rootJson, list, docAlloc);

	// Converts the document to text
	rapidjson::StringBuffer strBuf;
	rapidjson::PrettyWriter<rapidjson::StringBuffer> writer = rapidjson::PrettyWriter<rapidjson::StringBuffer>(strBuf);
	document.Accept(writer);
	jsonText = strBuf.GetString();
}

void JsonGenerator::getJson(rapidjson::Document::AllocatorType& docAlloc, const std::shared_ptr<db::table::Column> table, std::vector<rapidjson::Value>& jsonElements) {
	if (table == nullptr)
		return;
	// Get column name
	std::string columnName = table->getColumnName();

	// Get values
	std::vector<std::string> values;
	table->getValues(values);

	// Add elements to the JSON document
	size_t numRows = values.size();
	rapidjson::Value key(columnName.c_str(), docAlloc);
	for (size_t pos = 0; pos < numRows; ++pos) {
		rapidjson::Value key(columnName.c_str(), docAlloc);
		rapidjson::Value val(values.at(pos).c_str(), docAlloc);
		jsonElements.at(pos).AddMember(key,val, docAlloc);
	}

	// Add the neighbor column to the JSON document
	getJson(docAlloc, table->getNeighbor(), jsonElements);
}

void JsonGenerator::getJson(const std::string& root, const std::shared_ptr<db::table::Column> table, std::string& jsonText) {

	if (table == nullptr)
		return;

	// Create the document
	rapidjson::Document document;
	document.SetObject();
	rapidjson::Document::AllocatorType& docAlloc = document.GetAllocator();

	// Initialize the JSON document with the table structure
	rapidjson::Value list = rapidjson::Value(rapidjson::kArrayType);
	std::vector<std::string> column;
	table->getValues(column);
	size_t numRows = column.size();

	// Initialize the JSON document with the table content
	std::vector<rapidjson::Value> members = std::vector<rapidjson::Value>(numRows);
	for (rapidjson::Value& v : members) {
		v.SetObject();
	}
	getJson(docAlloc, table, members);

	for (rapidjson::Value& member : members) {
		rapidjson::Value key(root.c_str(), docAlloc);
		list.PushBack(member, docAlloc);
	}

	rapidjson::Value rootJson(root.c_str(), docAlloc);
	document.AddMember(rootJson, list, docAlloc);

	// Convert the JSON document to text
	rapidjson::StringBuffer strBuf;
	rapidjson::PrettyWriter<rapidjson::StringBuffer> writer = rapidjson::PrettyWriter<rapidjson::StringBuffer>(strBuf);
	document.Accept(writer);

	jsonText = strBuf.GetString();
}

} /* namespace rest */

File: java/src/main/java/rest/JSONGenerator.java
package main.java.rest;

import java.util.Vector;

import org.json.JSONArray;
import org.json.JSONObject;

import main.java.db.table.Column;

public class JSONGenerator {
	/**
	 * Create a JSON document given a list of values
	 * @param root			Root tag
	 * @param elements		List of values to generate the JSON document
	 * @return				JSON document in text format
	 */
	public static String getJson(final String root, final Vector<String> elements) {
		JSONObject jsonObject = new JSONObject();
		JSONArray jsonArray = new JSONArray(elements);
		jsonObject.put(root, jsonArray);

		// Return a formatted String
		return jsonObject.toString(2);
	}

	/**
	 * Create a JSON document given a database table
	 * @param root		Root tag
	 * @param table		Table values
	 * @return			JSON document in text format
	 */
	public static String getJson(final String root, final Column table) {	
		// Create a with the values and initialize it
		int size = table.getValues().size();
		Vector<JSONArray> elements = new Vector<JSONArray>(size);
		for (int i = 0; i<size; ++i) {
			elements.addElement(new JSONArray());
		}
		
		// Fill the vector elements with the elements of each Column
		// Each element of the vector corresponds to a row of the table
		Column row = table;
		Vector<String> values;
		do {
			values = row.getValues();
			int index = 0;
			// Go through all the values in the row and put them each in
			// a different array (list)
			for (String element : values) {
				JSONObject jsonObject = new JSONObject();
				jsonObject.put(row.getColumnName(), element);
				elements.elementAt(index).put(jsonObject);
				++index;
			}
		} while ((row = table.getNeighbor()) != null);
		
		// Create the root object with all the elements
		JSONObject jsonObject = new JSONObject();
		jsonObject.put(root, elements);
		
		// Return the formatted String
		return jsonObject.toString(2);
	}
}
